### PR TITLE
Cholmcc generators

### PR DESCRIPTION
**Pull request by cholmcc**
Please see the commit logs of the individual commits.

**TL;DR**: 

- `MCEventHeader`: Pre-defined keys for information
- `GeneratorPythia8`: Export _all_ available information to `MCEventHeader` 
- `GeneratorHepMC`:
  - Export _all_ available information to `MCEventHeader` 
  - Read events from external, spawned background program via FIFO 
- `GeneratorTParticle`: 
  - Reads from `TChain` with `TClonesArray` branch with `TParticle` objects 
    - Can read from existing files, _or_
    - from file produced by external, spawned background program

The changes to `GeneratorHepMC` allows us to use _any_ event generator that can write HepMC events. 

Examples of use are given in log messages and code comments. 

Yours,

_Christian_

### CODE REVIEW

**Comment by sawenzel** [comment]
Looks reasonable but would discuss this in the simulation meeting first of all.

**Comment by cholmcc** [comment]
Hi,

> Looks reasonable ... 

Great. 

> ... but would discuss this in the simulation meeting first of all.

OK, a few issues 

- Can you advice as to when such a meeting will take place?  Would you want my presence there? 
- This effort should be considered with the effort to integrate Rivet into the $\mathrm{O}^{2}$ pipeline.
  - That effort is taking place in the project [O2Rivet](https://gitlab.com/cholmcc/O2Rivet).   Please see there fore more on that project 
  - Rivet Analyses _may_ require more than just particles and vertexes, but also the auxiliary information typically present in the HepMC event header. 
- The idea of the changes to `GeneratorHepMC` is to allow integration of more event generators into the  $\mathrm{O}^{2}$ pipeline.  In particular, more and more EGs support HepMC.  I've integrated a number of EGs my self - f.ex. AMPT.

Yours,

_Christian_

File: Generators/include/Generators/GeneratorTParticle.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+///
+/// The tree name and particle branch names can be configured.
+///
+///   --configKeyValues "TParticle.treeName=T,TParticle.branchName=P"
+///
+/// File(s) to read are also configurable
+///
+///   --configKeyValues "TParticle.fileNames=foo.root,bar.root"
+///
+class GeneratorTParticle : public Generator
```
Comment (sawenzel) [inline Generators/include/Generators/GeneratorTParticle.h:51]: I am a bit unsure about this one: We already have a generator "GeneratorFromFile" which reads TParticles from a Run2 - like ROOT file.

A quick check if we still need this one would good.

File: Generators/include/Generators/GeneratorTParticleParam.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+   a parameter class/struct to keep the settings of the TGenerator
+   event generator and allow the user to modify them */
+struct GeneratorTParticleParam : public o2::conf::ConfigurableParamHelper<GeneratorTParticleParam> {
+  std::string treeName = "T";
+  std::string branchName = "Particles";
+  std::string fileNames = "tparticle.root";
+  std::string progCmd = "";
+  O2ParamDef(GeneratorTParticleParam, "TParticle");
```
Comment (sawenzel) [inline Generators/include/Generators/GeneratorTParticleParam.h]: "TParticle" as key sounds a bit short and might be confusing. Maybe we could go with "TParticleGen" or something (if we decide to keep GeneratorTParticle).

File: Generators/include/Generators/GeneratorTParticle.h (hunk 1 above)
Comment (cholmcc) [reply Generators/include/Generators/GeneratorTParticle.h:51]: There's a difference between the `GeneratorFromFile` and `GeneratorTParticle`, mainly in that `GeneratorTParticle` 
- is more flexible 
- can read data from external program rather than from file 
- can read from a `TChain` of files 
- assumes _all_ events are in the same `TTree`, unlike `GeneratorFromFile` which assumes there's one `TTree` for each event in the `TDirectory` named `Event<_event-no_>` in the single input file. 

As such, the two classes `GeneratorFromFile` and `GeneratorTParticle serves different needs 

- `GeneratorFromFile` is designed to read from an **AliROOT** `Kinematics.root` file
- `GeneratorTParticle` is designed to read from a general **ROOT** event tree, for example produced by a `TGenerator`. 

Hope that clarifies the differences.

File: Generators/include/Generators/GeneratorTParticleParam.h (hunk 2 above)
Comment (cholmcc) [reply Generators/include/Generators/GeneratorTParticleParam.h]: I used `TParticle` as the prefix for the keys to stay in line with other prefixes (e.g., `HepMC` for `GeneratorHepMC`, and `Pythia6` for `GeneratorPythia6`).  However, many of the generators key prefix are set to the name of the generator (e.g., `GeneratorPythia8` for `GeneratorPythia8`), so I think it would make more sense to use `GeneratorTParticle` if anything.  On the other hand `TParticleGen` I think is perhaps misleading, as it suggests that there's an actual generator process running, which may or may not be the case. 

Also since the MR, I've done some developments on `GeneratorTParticle` and `GeneratorHepMC` to derive from `GeneratorFileOrCmd` (in addition to deriving from `Generator`).   This base class has utilities for registering input file(s), set up a command line for a child program, FIFOs, temporary file names and so on.   Since the settings for that are common to all generators that derive from that (and `GeneratorFromFile` and `GeneratorO2Kine` could do that), I've introduced the class `GeneratorFileOrCmdParam` with prefix `FileOrCmd`.   That greatly simplifies the code and reduces code duplication quite a bit.   These changes are not yet anywhere to be found, as I'm waiting for this MR to go through.

**Comment by benedikt-voelkel** [comment]
Would it make sense to factorise the 3 things in this PR?
* `GeneratorTParticle`
* aligning `MCEventHeader` fields
* `HepMC` devs

In my opinion, your `HepMC` and `MCEventHeader` developments here might worth its own PR and should bring already a lot of flexibility you might need for your `O2Rivet` developments and tests.
In fact, it gives users immediately the possibility to play with other (very custom generators).

Could you remind me maybe what the exact reason for having the `GeneratorTParticle`? Maybe I missed that in your presentation yesterday. Is it a blocker for `O2Rivet` developments or could they go forward at this point without another generator type?

After that, another PR could introduce further O2 generator developments.

File: Generators/include/Generators/GeneratorTParticle.h (hunk 1 above)
Comment (cholmcc) [reply Generators/include/Generators/GeneratorTParticle.h:51]: Oh, and another difference.   Using the key `progCmd`, one could execute a `TGenerator` as a background process.  Say we have the script `MyEG.C`

```
class MyGenerator : public TGenerator 
{
   ...
};
 
void MyEG(const char* output, Int_t nEv, Int_t seed, Float_t bmax)
{
   gRandom->SetSeed(seed);
  
   MyGenerator generator(bMax);
  TFile* file = TFile::Open(output, "RECREATE");
  TTree* tree = new TTree("T");
  TClonesArray* particles = new TClonesArray("TParticle");
  tree->Branch("Particles", &particles);
  tree->SetDirectory(file);

  for (Int_t iEv = 0; iEv < nEv; iEv++) {
     generator.GenerateEvent();
     generator.ImportParticles(particles);
     tree->Fill();
     tree->AutoSave("SaveSelf FlushBaskets Overwrite");
  }
  file->Write();
  file->Close();
}
``` 

and a simple script like `myeg.sh` 

```
#!/bin/sh 

nev=1
out=myeg.root 
seed=0
bmax=0

while test $# -gt 0; do 
   case $1 in 
   -n|--nevents) nev=$2 ; shift ;;
   -o|--output)   out=$2 ; shift ;;
   -b|--bMax)     bmax=$2; shift;;  # Future proof 
   -s|--seed)       seed=$2; shift;;
  --help) echo "Usage: $0 [-n NEV] [-o OUTPUT] [-s SEED] [-b bmax]"; exit 0;;
  esac 
root MyEG.C++\( \"$out\",$nev,$seed,$bmax\)
```

then one can do 

```
o2-sim -g tparticle --configKeyValues "TParticle.progCmd=myeg.sh"
```

which means all the old `AliGenerators` can be used more or less directly :-)

**Comment by cholmcc** [comment]
Hi Sandro, 

I missed your last comment.  Perhaps the point raised [above](https://github.com/AliceO2Group/AliceO2/pull/11913#discussion_r1340132394) could be enough to convince you that `GeneratorTParticle` should also go in #11699 

`GeneratorTParticle` is not a stumbling block - in any way - for me, just another flexible option for everyone to play with. 

If you insist on taking out `GeneratorTParticle`, could you please advice how to best take that out of the MR.  I suspect doing `git rm` to the files and some other edits and then a new commit to the MR branch, but if you know of a better way, please let me know.  Thanks.

**Comment by cholmcc** [comment]
Hi Sandro, 

Any news on this?   Please advice as to what you see as the way forward (vis-a-vis the discussion here and on MatterMost).  Thanks.

**Comment by sawenzel** [comment]
I don't mind merging this as a whole and there is no problem introducing the `GeneratorTParticle` class (the existing one is indeed specific to AliRoot). 

However, I'd like to insist that
(a) the configurable key is not just called "TParticle". It is just not clear from this name, that it is configuring a generator.

and 

(b) while you are modifying the PR, please also add the usage examples (for HepMC, GeneratorTParticle) in `/run/SimExamples`.

**Review by sawenzel — DISMISSED** [review]
